gapis/gfxapi/vulkan/vulkan_terminator.go

```diff
@@ -0,0 +1,378 @@
+		NewVoidᶜᵖ(memory.Nullptr),
+		commandBuffer.Pool,
+		VkCommandBufferLevel_VK_COMMAND_BUFFER_LEVEL_PRIMARY,
+		uint32(1),
+	}
+	allocateData := atom.Must(atom.AllocData(ctx, s, allocate))
+	commandBufferData := atom.Must(atom.AllocData(ctx, s, commandBufferId))
+
+	x = append(x,
+		NewVkAllocateCommandBuffers(commandBuffer.Device,
+			allocateData.Ptr(), commandBufferData.Ptr(), VkResult_VK_SUCCESS,
+		).AddRead(allocateData.Data()).AddWrite(commandBufferData.Data()))
+
+	beginInfo := VkCommandBufferBeginInfo{
+		VkStructureType_VK_STRUCTURE_TYPE_COMMAND_BUFFER_BEGIN_INFO,
+		NewVoidᶜᵖ(memory.Nullptr),
+		VkCommandBufferUsageFlags(VkCommandBufferUsageFlagBits_VK_COMMAND_BUFFER_USAGE_ONE_TIME_SUBMIT_BIT),
+		NewVkCommandBufferInheritanceInfoᶜᵖ(memory.Nullptr),
+	}
+
+	beginInfoData := atom.Must(atom.AllocData(ctx, s, beginInfo))
+	x = append(x,
+		NewVkBeginCommandBuffer(commandBufferId, beginInfoData.Ptr(), VkResult_VK_SUCCESS).AddRead(beginInfoData.Data()))
+
+	// DestroyResourcesAtEndOfFrame will handle this actually removing the
+	// command buffer. We have no way to handle WHEN this will be done
+	for i := 0; i < int(writeIdx)+additional; i++ {
+		cmd := commandBuffer.Commands[i]
+		c, a := AddCommand(ctx, commandBufferId, s, cmd.recreateData)
+		x = append(x, a)
+		cleanup = append(cleanup, c)
+	}
+	for i := range additionalCommands {
+		c, a := AddCommand(ctx, commandBufferId, s, additionalCommands[i])
+		x = append(x, a)
+		cleanup = append(cleanup, c)
+	}
+	x = append(x,
+		NewVkEndCommandBuffer(commandBufferId, VkResult_VK_SUCCESS))
+	cleanup = append(cleanup, func() {
+		allocateData.Free()
+		commandBufferData.Free()
+		beginInfoData.Free()
+	})
+	return VkCommandBuffer(commandBufferId), x, cleanup
+}
+
+func CutCommandBuffer(ctx context.Context, id atom.ID,
+	a atom.Atom, idx gfxapi.SubcommandIndex, out transform.Writer) {
+	submit := a.(*VkQueueSubmit)
+	s := out.State()
+	c := GetState(s)
+	l := s.MemoryLayout
+	o := a.Extras().Observations()
+	o.ApplyReads(s.Memory[memory.ApplicationPool])
+	submitInfo := submit.PSubmits.Slice(uint64(0), uint64(submit.SubmitCount), l)
+	_ = submitInfo
+	skipAll := len(idx) == 0
+
+	// Notes:
+	// - We should walk/finish all unfinished render passes
+	// - if cutLastElement, then ignore the last element that we would run
+	// idx[0] is the submission index
+	// idx[1] is the primary command-buffer index in the submission
+	// idx[2] is the command index in the primary command-buffer
+	// idx[3] is the secondary command buffer index inside a vkCmdExecuteCommands
+	// idx[4] is the secondary command inside the secondary command-buffer
+	submitCopy := NewVkQueueSubmit(submit.Queue, submit.SubmitCount, submit.PSubmits,
+		submit.Fence, submit.Result)
+	submitCopy.Extras().Add(a.Extras().All()...)
+
+	newCommandBuffers := make([]VkCommandBuffer, 1)
+	lastSubmit := uint64(0)
+	lastCommandBuffer := uint64(0)
+	if !skipAll {
+		lastSubmit = idx[0]
+		lastCommandBuffer = idx[1]
```

Comment:
I agree creating a new command buffer in all cases. But if `len(idx) == 1`, isn't `idx[1]` will crash?